### PR TITLE
Overlap grammar (constrained decoding) with speculative decode verify

### PR DESCRIPTION
**Pull request by merrymercy**
Supersedes #14169 and #15623 (earlier grammar + spec-decode overlap explorations).

## Motivation

When speculative decoding runs together with grammar (constrained decoding), grammar-constrained spec decode needs a CPU round trip every verify step: copy the draft tokens out (D2H), advance the grammar FSM over the previous step's committed tokens, build the vocab bitmask, and copy it back (H2D). Today these run *around* the GPU instead of *under* it, and the scheduler also forces cross-batch overlap off, so every grammar decode step has large GPU bubbles:

```text
CURRENT (main) — grammar + spec decode, one step N

CPU: [ accept_grammar(N-1) + detok(N-1) ]   [ D2H draft(N) ]        [ bitmask(N) ][ H2D mask ]
       pop_and_process, overlap OFF           blocking .cpu()         build        pageable
                                                                                   = blocking

GPU: ####### idle #######  ..draft(N)..  ## idle ##  ==== target verify(N) ====  ## idle ##  sample(N)
     ^ N-1 must be fully                  ^ waits for draft;                      ^ waits for
       processed before N can launch        target not launched yet                blocking H2D
       (need_grammar_sync)
```

Two root causes:
1. **Worker path** — the copies are blocking and mis-ordered: the draft-token `.cpu()` is issued *before* the target forward is launched (the GPU drains), and the mask H2D copies from *pageable* host memory (so `non_blocking=True` is a silent no-op).
2. **Scheduler path** — `need_grammar_sync` disables overlap for every grammar + spec decode batch, forcing the previous batch's result processing (which includes the FSM advance) to run before the next batch launches.

## Approach

The only consumer of the advanced grammar FSM is `generate_token_bitmask(N)`, which runs *after* the target-verify forward is launched. So we move every piece of grammar CPU work into that window — the FSM advance for N-1 and the bitmask build for N both run while the GPU executes verify(N):

```text
PROPOSED — grammar + spec decode, one step N (overlap ON)

CPU: launch(N) -> accept_grammar(N-1) -> generate_bitmask(N) -> H2D(pinned, async) -> sample(N)
                  \___________ all grammar CPU work overlaps the GPU forward ___________/

GPU: ..draft(N).. ================= target verify(N) forward =================  sample(N)

     (detok / output(N-1) also overlap — result processing now runs after the launch)
```

`draft(N)` can safely run before `accept_grammar(N-1)`: the draft model is grammar-unconstrained (grammar only masks the *target* logits at verify), and it consumes only N-1's committed token *values* via the GPU spec relay (`prev_target_prediction` / bonus tokens) — never the FSM. The FSM advance is sequenced only ahead of its real consumer, `generate_token_bitmask(N)`.

## Modifications

- **Capability flag**: add `SpeculativeAlgorithm.supports_grammar_overlap()` (true for EAGLE) and a `False` default on `CustomSpecAlgo`. When true, the scheduler hands the worker a grammar barrier via `forward_batch_generation` instead of disabling overlap; when false, the previous "force overlap off" behavior is preserved.
- **Scheduler (keep overlap on)**: `_advance_pending_grammar()` advances the FSM over any decode result still pending in the `result_queue`; it is passed as `grammar_barrier` so the CPU advance overlaps the target verify forward, and it also resolves the extend→decode boundary. `need_grammar_sync` is now skipped for grammar-overlap-capable algorithms so the previous batch's remaining result processing overlaps as usual.
- **Single FSM advance point**: consolidate the advance into an idempotent `SchedulerBatchResultProcessor.advance_grammar_fsm()`, handling both decode (the accepted spec run, memoized as `grammar_retained_tokens` so `pop_and_process(N-1)` reuses them with no double-advance) and extend (the single prefilled token). It runs either eagerly via the barrier (inside `verify()`) or lazily from the result processors; `GenerationBatchResult` tracks `grammar_advanced` / `grammar_retained_tokens`.
- **`run_eagle_verify` (overlap the copies with the forward)**: call the grammar barrier right before the bitmask build, and issue the draft / verify-input D2H copies asynchronously (pinned buffers + a recorded CUDA event synced just before the bitmask build) so the FSM advance and the CPU bitmask traversal overlap the target verify forward. Use a `non_blocking` H2D for the vocab mask. `grammar_barrier` is threaded through both EAGLE workers (`eagle_worker_v2` and `multi_layer_eagle_worker_v2`).
- **xgrammar (pinned bitmask)**: allocate the token bitmask as a pinned tensor (`_allocate_token_bitmask`) so the subsequent device copy is a genuine `non_blocking` H2D (a pageable source silently downgrades it to a blocking copy).
- **ScheduleBatch.copy()**: carry `has_grammar` so queued batch copies self-gate in `advance_grammar_fsm`.

## Notes for reviewers

The upstream EAGLE verify path (`run_eagle_verify` in `eagle_worker_common.py`) differs from the code this change originated on, where the grammar handling was inline in the worker's `verify()`. The change has been adapted onto the shared `run_eagle_verify` and threaded through its two callers. The concurrency-sensitive paths (async D2H + grammar barrier ordering relative to the target forward) were not exercised on GPU in this environment; CI and reviewers on the EAGLE + grammar path should validate.

## Checklist

- Verified the non-GPU pieces compile.
- CI to validate the EAGLE + grammar overlap path.

## Original commits

- \`a7d71bfa7\`

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #29633233288](https://github.com/sgl-project/sglang/actions/runs/29633233288)
Latest PR Test (Extra): :x: [Run #29633233194](https://github.com/sgl-project/sglang/actions/runs/29633233194)

### CODE REVIEW

**Comment by merrymercy** [comment]
/rerun-test test/registered/constrained_decoding/test_constrained_decoding.py test/registered/openai_server/features/test_json_mode.py test/registered/spec/eagle/test_eagle_constrained_decoding.py test/registered/unit/constrained/test_base_grammar_backend.py test/registered/unit/constrained/test_grammar_manager.py test/registered/unit/managers/test_batch_result_processor_spec_grammar.py test/registered/unit/spec/test_spec_utils_traverse_tree.py

**Comment by merrymercy** [comment]
/rerun-test test/registered/constrained_decoding/test_constrained_decoding.py test/registered/openai_server/features/test_json_mode.py test/registered/spec/eagle/test_eagle_constrained_decoding.py test/registered/unit/constrained/test_base_grammar_backend.py test/registered/unit/constrained/test_grammar_manager.py test/registered/unit/managers/test_batch_result_processor_spec_grammar.py test/registered/unit/spec/test_spec_utils_traverse_tree.py

**Comment by merrymercy** [comment]
/rerun-test test/registered/constrained_decoding/test_constrained_decoding.py test/registered/openai_server/features/test_json_mode.py test/registered/spec/eagle/test_eagle_constrained_decoding.py test/registered/unit/constrained/test_base_grammar_backend.py test/registered/unit/constrained/test_grammar_manager.py test/registered/unit/managers/test_batch_result_processor_spec_grammar.py test/registered/unit/spec/test_spec_utils_traverse_tree.py

**Comment by merrymercy** [comment]
/rerun-test test/registered/constrained_decoding/test_constrained_decoding.py test/registered/openai_server/features/test_json_mode.py test/registered/spec/eagle/test_eagle_constrained_decoding.py test/registered/spec/test_frozen_kv_mtp.py test/registered/unit/constrained/test_base_grammar_backend.py test/registered/unit/constrained/test_grammar_manager.py test/registered/unit/managers/test_batch_result_processor_spec_grammar.py test/registered/unit/spec/test_spec_utils_traverse_tree.py

**Review by hnyls2002 — APPROVED** [review]
